### PR TITLE
`booth` cleanup

### PR DESCRIPTION
**Pull request by povik**
Following up to #3924, this PR will to some degree clean up the code of the Booth pass.

### CODE REVIEW

**Comment by povik** [comment]
@andyfox-rushc Take a look, feel free to ask and/or object.

File: passes/techmap/booth.cc [hunk 1]
```diff
@@ -797,32 +764,23 @@ struct BoothPassWorker {
...
-
 		static int cpa_id;
 		cpa_id++;
 
-		RTLIL::Wire *carry = nullptr;
-
-		log_assert(s_vec.size() == c_vec.size());
+		log_assert(c_vec.size() == s_vec.size());
+		// TODO: doesn't pass
+		//log_assert(result.size() == s_vec.size() + 2);
```
Comment (povik) [inline passes/techmap/booth.cc:659]: @andyfox-rushc Here I moved the `result.size() == s_vec.size() + 2` condition down from the "axioms" comment to an actual assert, but it wasn't passing. Should this be upheld?

Comment (andyfox-rushc) [reply passes/techmap/booth.cc:659]: Hi, Martin,
My comment is wrong and the axiom you refer to does not need to be upheld. The axiom for the unsigned case should be result.size() ==s_vec.size(). 
In this unsigned case an array of CSA (Wallace) trees is used: 1 per result bit.  Each CSA tree produces a sum and carry. These are summed up by the CPA. The final carry out is ignored in the unsigned case. My error comes from the use of the same structure for signed multiplication, in which case the multiplier is not padded out and one partial product removed which leads to the different axiom for the size. But again my comment was wrong for the unsigned case.

Comment (povik) [reply passes/techmap/booth.cc:659]: Thanks for the analysis! I wanted to make sure there's not an issue lurking there.

Comment (andyfox-rushc) [reply passes/techmap/booth.cc]: PS Martin, sorry for being quiet. I did quickly go through all your changes which look good to me. I like the addition of the full adder helper creator to RTLIL.  The simplifications to the cell selection and the extension of the arguments to fit the right size (I was adding weird buffers), all seem to make the code much clearer to me. I see these all got safely merged in today. All looks great to me. Thanks.

**Pull request merged**